default/python/AI/ProductionAI.py

```diff
@@ -1611,59 +1272,57 @@ def already_has_completed_colony_building(planet_id) -> bool:
     return any(universe.getBuilding(bldg).name.startswith("BLD_COL_") for bldg in planet.buildingIDs)
 
 
-def build_ship_facilities(bld_name, top_locs=None):
+def _build_ship_facilities(building_type: BuildingType, top_pids: Set[PlanetId] = None) -> None:
     # TODO: add total_pp checks below, so don't overload queue
-    if top_locs is None:
-        top_locs = set()
+    if top_pids is None:
+        top_pids = set()
     universe = fo.getUniverse()
     empire = fo.getEmpire()
     total_pp = empire.productionPoints
-    __, prereq_bldg, this_cost, time = AIDependencies.SHIP_FACILITIES.get(bld_name, (None, "", -1, -1))
-    if not get_aistate().character.may_build_building(bld_name):
-        return
-    bld_type = fo.getBuildingType(bld_name)
-    if not empire.buildingTypeAvailable(bld_name):
+    prerequisite_type = building_type.prerequisite()
+    if not building_type.available():
         return
-    queued_bld_locs = [element.locationID for element in empire.productionQueue if element.name == bld_name]
-    if bld_name in AIDependencies.SYSTEM_SHIP_FACILITIES:
-        current_locs = get_systems_with_facilities(bld_name)
-        current_coverage = current_locs.union(universe.getPlanet(planet_id).systemID for planet_id in queued_bld_locs)
+    queued_bld_pids = building_type.queued_at()
+    if building_type in BuildingType.get_system_ship_facilities():
+        current_coverage = building_type.built_or_queued_at_sys()
         open_systems = set(
-            universe.getPlanet(pid).systemID for pid in get_best_pilot_facilities("BLD_SHIPYARD_BASE")
+            universe.getPlanet(pid).systemID for pid in BuildingType.SHIPYARD_BASE.get_best_pilot_facilities()
         ).difference(current_coverage)
-        try_systems = (
-            open_systems.intersection(get_systems_with_facilities(prereq_bldg)) if prereq_bldg else open_systems
-        )
-        try_locs = set(pid for sys_id in try_systems for pid in get_owned_planets_in_system(sys_id))
+        try_systems = open_systems & prerequisite_type.built_or_queued_at_sys() if prerequisite_type else open_systems
+        try_pids = set(pid for sys_id in try_systems for pid in get_owned_planets_in_system(sys_id))
```

Comment:
You could use set comprehansions

```suggestion
        try_pids = {pid for sys_id in try_systems for pid in get_owned_planets_in_system(sys_id)}
```

Reply:
Yes, I guess I can. But that is what meant with one step at a time:
My only change in that line was to change the name try_locs to try_pids, because the existing code used _locs for both planets and systems.